api/pkg/s3/objectput.go

```diff
@@ -15,92 +15,188 @@
-		object.Backend = backendName
-		client = getBackendByName(ctx, s, backendName)
+	if size == -1 {
+		response.WriteError(http.StatusLengthRequired, s3error.ErrMissingContentLength)
+		return
+	}
+
+	// maximum Upload size for objects in a single operation
+	if isMaxObjectSize(size) {
+		response.WriteError(http.StatusBadRequest, s3error.ErrEntityTooLarge)
+		return
+	}
+
+	// Save metadata.
+	metadata := extractMetadataFromHeader(request)
+	// Get Content-Md5 sent by client and verify if valid
+	if _, ok := request.Request.Header["Content-Md5"]; !ok {
+		metadata["md5Sum"] = ""
 	} else {
-		bucket, _ := s.s3Client.GetBucket(ctx, &s3.Bucket{Name: bucketName})
-		object.Backend = bucket.Backend
-		client = getBackendClient(ctx, s, bucketName)
+		if len(request.Request.Header.Get("Content-Md5")) == 0 {
+			log.Infoln("Content Md5 is null!")
+			response.WriteError(http.StatusBadRequest, s3error.ErrInvalidDigest)
+			return
+		}
+		md5Bytes, err := checkValidMD5(request.Request.Header.Get("Content-Md5"))
+		if err != nil {
+			log.Infoln("Content Md5 is invalid!")
+			response.WriteError(http.StatusBadRequest, s3error.ErrInvalidDigest)
+			return
+		} else {
+			metadata["md5Sum"] = hex.EncodeToString(md5Bytes)
+		}
 	}
 
-	if client == nil {
-		response.WriteError(http.StatusInternalServerError, NoSuchBackend.Error())
+	acl, err := getAclFromHeader(request)
+	if err != nil {
+		WriteErrorResponse(request, response, err)
 		return
 	}
-	log.Infof("enter the PUT method")
-	s3err := client.PUT(request.Request.Body, &object, ctx)
-	log.Infof("LastModified is %v\n", object.LastModified)
-	if s3err != NoError {
-		response.WriteError(http.StatusInternalServerError, s3err.Error())
+
+	// check if specific bucket exist
+	ctx := common.InitCtxWithAuthInfo(request)
+	bucketMeta := s.getBucketMeta(ctx, bucketName)
+	if bucketMeta == nil {
+		response.WriteError(http.StatusInternalServerError, s3error.ErrGetBucketFailed)
 		return
 	}
+	//log.Logf("bucket, acl:%f", bucketMeta.Acl.CannedAcl)
+	location := bucketMeta.DefaultLocation
+	if backendName != "" {
+		// check if backend exist
+		if s.isBackendExist(ctx, backendName) == false {
+			response.WriteError(http.StatusBadRequest, s3error.ErrGetBackendFailed)
+		}
+		location = backendName
+	}
 
-	res, err := s.s3Client.CreateObject(ctx, &object)
+	var limitedDataReader io.Reader
+	if size > 0 { // request.ContentLength is -1 if length is unknown
+		limitedDataReader = io.LimitReader(request.Request.Body, size)
+	} else {
+		limitedDataReader = request.Request.Body
+	}
+
+	buf := make([]byte, ChunkSize)
+	eof := false
+	stream, err := s.s3Client.PutObject(ctx)
+	defer stream.Close()
+	obj := pb.Object{
+		BucketName: bucketName,
+		ObjectKey: objectKey,
+		Acl: &pb.Acl{CannedAcl: acl.CannedAcl},
+		CustomAttributes:metadata,
+		Location: location,
+		Size: size,
+	}
+	err = stream.SendMsg(&obj)
 	if err != nil {
-		log.Errorf("err is %v\n", err)
-		response.WriteError(http.StatusInternalServerError, err)
```

Comment:
how do we distinguish between
1. internal server error in OSDS...our end
2. actual errors on the vendor backend like AWS etc. those need to be reported verbatim

Reply:
@rhsakarpos, The errors database return and vendor backend return are all recorded in log,  and client will received http.StatusInternalServerError at the similar cases,  and we can check the log to fix http.StatusInternalServerError.  What about your suggestion?

Reply:
my suggestion is to pass on cloud vendor errors as is...to the UI. InternalServerError means error in OSDS server, no ?

Reply:
@rhsakarpos ,  Cloud vendor errors are passed to UI with errors's string description, but error code is still InternalServerError.  This is my idea. If you agree with me, I will add it to issue, and modify it at next PR.  Thanks